### PR TITLE
[clang] Output an error when [[lifetimebound]] attribute is applied on a function parameter while the function returns void

### LINKED ISSUE #107556: [clang] Clang must diagnose the use of [[lifetimebound]] annotation in void-returning functions
https://godbolt.org/z/rcPnrWrr7

```cpp
#include <string>
struct S {
    void setS(const std::string& s [[clang::lifetimebound]]) { this->s = &s; }
    const std::string* s;
};

void foo() {
    S s;
    s.setS(std::string());
}
```

Lifetime analysis is currently not defined for parameters annotated with `[[lifetimebound]]` in void-returning function.
Current analysis only enforces that the annotated parameter outlives the **return** value. With no return value, this analysis is silently disabled.

### PR DESCRIPTION
**Pull request by bricknerb**
Fixes: https://github.com/llvm/llvm-project/issues/107556

### CODE REVIEW

File: clang/lib/Sema/SemaDecl.cpp [hunk 1]
```diff
@@ -6970,6 +6970,18 @@ static void checkAttributesAfterMerging(Sema &S, NamedDecl &ND) {
         }
       }
     }
+
+    for (unsigned int I = 0; I < FD->getNumParams(); ++I) {
+      const ParmVarDecl *P = FD->getParamDecl(I);
+
+      // The [[lifetimebound]] attribute can be applied to a function parameter
+      // only if the function returns a value.
+      if (auto *A = P->getAttr<LifetimeBoundAttr>()) {
```
Comment (usx95) [inline clang/lib/Sema/SemaDecl.cpp:6979]: An interesting case that [came up](https://github.com/llvm/llvm-project/issues/107556#issuecomment-2333906528) involves dependent return type.

I think we should not be showing this error if there is a dependent return type.
```cpp
template <class T, class R = void> R foo(const T& t [[clang::lifetimebound]]);
foo<int>(1);
const int& r = foo<int, int>(1);
```

There is no way to fix this for one instantiation and not for the other so we should be accepting these.

Can you also add test involving tempates with and without dependent return types.

File: clang/lib/Sema/SemaDecl.cpp [hunk 2]
```diff
@@ -6970,6 +6970,18 @@ static void checkAttributesAfterMerging(Sema &S, NamedDecl &ND) {
         }
       }
     }
+
+    for (unsigned int I = 0; I < FD->getNumParams(); ++I) {
```
Comment (usx95) [inline clang/lib/Sema/SemaDecl.cpp:6974]: We would also want to diagnose this for member functions where lifetimebound is not attached to a param but to the function type.
```cpp
struct S {
  void member() [[clang::lifetimebound]];
}
```

File: clang/lib/Sema/SemaDecl.cpp (hunk 1 above)
Comment (bricknerb) [reply clang/lib/Sema/SemaDecl.cpp:6979]: I believe we don't.
I've checked and the return type is TemplateTypeParm so it is not considered to be void.

Comment (cor3ntin) [reply clang/lib/Sema/SemaDecl.cpp:6979]: I think if the type is dependent, the check should be skipped.
Hopefully the check is then run upon instantiation

Comment (bricknerb) [reply clang/lib/Sema/SemaDecl.cpp:6979]: Added a test to demonstrate that.

File: clang/lib/Sema/SemaDecl.cpp (hunk 2 above)
Comment (bricknerb) [reply clang/lib/Sema/SemaDecl.cpp:6974]: Makes sense!
Added a test for that with FIXME to fix in another pull request.

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -133,6 +133,15 @@ C++ Specific Potentially Breaking Changes
     // Fixed version:
     unsigned operator""_udl_name(unsigned long long);
 
+- Clang will now produce an error diagnostic when [[clang::lifetimebound]] is
+  applied on a parameter of a function that returns void. This was previously 
+  gnored and had no effect.
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:138]:
```suggestion
  gnored and had no effect. (#GH107556)
```

**Review by cor3ntin — APPROVED** [review]
LGTM modulo comment. Thanks!

File: clang/docs/ReleaseNotes.rst (hunk 3 above)
Comment (bricknerb) [reply clang/docs/ReleaseNotes.rst]: Done.

**Review by usx95 — APPROVED** [review]
LGTM.

File: clang/test/SemaCXX/attr-lifetimebound.cpp [hunk 4]
```diff
@@ -31,6 +32,11 @@ namespace usage_ok {
     return *(int*)param;
   }
 
+  template <class T, class R = void> R dependent_void(const T& t [[clang::lifetimebound]]);
+  void dependent_void_instantiation() {
+    dependent_void<int>(1);
```
Comment (usx95) [inline clang/test/SemaCXX/attr-lifetimebound.cpp:37]: nit: add the following for completeness.
```
int x = dependent_void<int, int>(1); // warning.
dependent_void<int, int>(1); // Ok.
```

**Review by usx95 — COMMENTED** [review]
small nit for one more test.

File: clang/test/SemaCXX/attr-lifetimebound.cpp (hunk 4 above)
Comment (bricknerb) [reply clang/test/SemaCXX/attr-lifetimebound.cpp]: Done.

**Pull request merged**